### ISSUE TITLE
Inconsistent null parameter handling in IEqualityComparer.GetHashCode(Object)

### ISSUE THREAD

**Issue by ohadschn** [dotnet/corefx#28170]
The [docs](https://docs.microsoft.com/en-us/dotnet/api/system.collections.iequalitycomparer.gethashcode?view=netframework-4.7.1#System_Collections_IEqualityComparer_GetHashCode_System_Object_) state that `IEqualityComparer.GetHashCode(Object)` should throw `ArgumentNullException` when "The type of `obj` is a reference type and `obj` is `null`".

However, this contract is inconsistently implemented across the framework:
1. `null` parameter leads to `ArgumentNullException` as expected: https://github.com/dotnet/corefx/blob/2b805ce320feab0d7d9b72123767cc017c32948f/src/Common/src/CoreLib/System/StringComparer.cs#L305
1. `null` parameter results in no exception and returns 0: https://github.com/dotnet/corefx/blob/6dd451f51451a7d0ceea6104b51bd17005e9a0e6/src/System.Collections/src/System/Collections/Generic/HashSetEqualityComparer.cs#L34
1. `null` parameter is not checked at all, leading to `NullReferenceException`: https://github.com/dotnet/corefx/blob/1514f1cb419ada9461730b05c1b67caef4aba0e2/src/System.Private.Xml/src/System/Xml/Core/SecureStringHasher.cs#L28

**Comment by karelz** [comment]
IMO those are sad mistakes of the past. Changing the implementation would mean breaking code which already depends on current exceptions. Also it would mean additional differences between .NET Framework and .NET Core. While the exception is not perfect, at least it is an exception. I would suggest to not fix these differences, maybe just document them properly. I let area experts to make the final call.

**Comment by ohadschn** [comment]
@karelz Thank you. So just to clarify, the correct behavior (when implementing new classes) is what the docs say, right? Namely, check for nullity and throw `ArgumentNullException` correspondigly?

**Comment by karelz** [comment]
@ohadschn yes, the docs should be viewed as recommended way to implement it.
The cases [2] and [3] seem to be internal. Do you know if they expose their non-standard behavior in public APIs?

**Comment by ohadschn** [comment]
Well you can get the comparer from case [2] using the public `HashSet<int>.CreateSetComparer()`.
I couldn't find any public example for case [3].

**Comment by akarpov89** [comment]
@karelz  @ohadschn There are way more implementations (in BCL and in 3rd party libraries) which don't throw an `ArgumentNullException` in case argument is `null`. And that's kind of ok in my opinion. But let's first get back to the docs and what they really mean.

The intention is clearly to set a *precondition*, which is in case of `IEqualityComparer<T>.GetHashCode` is `obj != null`. The responsibility to meet the preconditions lies on caller, i.e. it's a programmer's error to pass the `null`. But what should the callee do? The callee's responsibility is usually to ensure postconditions and maintain invariants (if any.) But with regard to precondition violation there are not many options to do.

First option is to just ignore the fact that precondition wasn't met and continue the execution, which will eventually lead to some runtime exception (e.g. `NullReferenceException`, `IndexOutOfRangeException` etc.) But this approach could make finding the bug harder.

The second option is to throw an exception. Which one? Basically it doesn't matter whether you throw an `ArgumentNullException` or some other `AssertionException` (like the one defined in Code Contracts library.) The goal of such exception is to break the control flow as earlier as possible to make debugging easier. The advantage of `AssertionException` is that it's private, i.e. you can't catch it explicitly (yet you still can catch general `Exception`), but you definitely can catch `ArgumentNullException` and continue execution (which is a terrible idea.)

The third option is to terminate the execution altogether (preferably after logging the error.) Unfortunately at some point engineers implementing BCL were not brave enough to use this mechanism to distinguish between recoverable errors and mere bugs.

Now let's get back to `IEqualityComparer<T>.GetHashCode`. Why am I saying that having hundreds (maybe even thousands) of implementations not checking argument for null is kind of ok? Because BCL collections (as primary consumers of `IEqualityComparer<T>` interface) meet the precondition and wouldn't call custom comparers with `null` argument.

Now regarding the `HashSet<T>`. The `HashSet<T>`, unlike `Dictionary<TKey, TValue>`, supports `null` values as "keys", i.e. you can add `null` into the set and it's by design. That's why its comparer doesn't throw an exception. Does it mean that `HashSetEqualityComparer` violates the contract? Not in any way! Preconditions can be weakened (and postconditions can be strengthened), it's absolutely ok and doesn't break any client code.
